### PR TITLE
Minor change to the internal structure of CAN packet implementation

### PR DESCRIPTION
**Pull request by polybassa**
I need this modifications to have a fixed packet size of 16 bytes. In my SocketCAN implementation I need to do a byteswap, since the Linux CAN sockets use a different byteorder as pcap.

### CODE REVIEW

**Comment by guedou** [comment]
@smainand what’s your opinion ? Can’t we use your byte swap feature ?

**Comment by smainand** [comment]
Indeed the swap-byte option was added for this purpose.
@polybassa does adding ``` conf.contribs['CAN']['swap-bytes'] = True ``` at the top of your script work for you ? Do you have an example pcap otherwise ?

**Comment by polybassa** [comment]
Hi, swap-bytes will do the job for me. I will be able to generate correct CAN-Frames in my socket. 

There is one thing, I don't like. If I do the following thing:

```
conf.contribs['CAN']['swap-bytes'] = True
rx = CAN(identifier=0x7ff,length=8,data=b'\x01\x02\x03\x04\x05\x06\x07\x08')
wrpcap('/tmp/scapyPcapTest1.pcap', rx, append=False)

conf.contribs['CAN']['swap-bytes'] = False
rx = CAN(identifier=0x7ff,length=8,data=b'\x01\x02\x03\x04\x05\x06\x07\x08')
wrpcap('/tmp/scapyPcapTest2.pcap', rx, append=False)
```
I will end up with two different pcap files. 
When I receive bytes on my socket, I've to set ```swap-bytes```. If I want to write the captured CAN Packets to a pcap file I've to reset  ```swap-bytes```. Otherwise I end up with wrong identifiers and flags in the pcap file. This will lead to messy code and errors during the usage of CANSockets. Therefore I would like to directly swap bytes in my Socket. Through this change, I can work with the pcap byteorder inside scapy. 
My changes modify the Field_desc of CAN(Packet) to ensure, always having at least 16 Bytes.

**Comment by smainand** [comment]
I find it convenient to be able to choose how the packets will be encoded to the pcap file. It allows for example to simply convert a pcap file from one format to another. 

I have not encountered a case where the packet size was a problem, but if you did adding the PadFields looks like a good solution. @guedou what do you think ?

Out of curiosity, what do you use the ```CAN_FRAME_SIZE``` and ```CAN_INV_FILTER``` for ?

**Comment by polybassa** [comment]
I'm planning to add CANSockets, ISOTPSockets, DoIPSockets and protocol definitions for UDS to scapy. 
I'll use this defines in my socket implementation. I can also move this defines inside the socket definition, but I think the fit better into the CAN Packet definition file. 
This PRs (which I have to redo) show my automotive stuff, that I want to add to scapy. #1205 #1208 #1209

Having the PadField would guarantee me a minimal size of 16 bytes, which help me in my implementation. 
Which byte order would you prefeer inside scapy? The natural Linux CAN-Socket byteorder, or the pcap byteorder? For usability, it would be very helpful, to swap bytes in my socket implementation. What do you think about that?

**Comment by smainand** [comment]
That's great news that you are working on the CANsocket implementation. Indeed in this case it makes more sense to perform the swap directly in it.

I guess it would be more natural to keep the pcap byteorder inside scapy ?

**Comment by polybassa** [comment]
Yes, I'll will do that. The byteswap option can still be present. I'll test how that option will influence my socket implementation. Maybe I'll have to make my internal byteswap inside the socket also dependant on this option.

**Comment by polybassa** [comment]
As soon as this PR is merged, I'll can continue with my CANSocket PR.

File: scapy/layers/can.py [hunk 1]
```diff
@@ -24,6 +23,8 @@
 #   set to True when working with PF_CAN sockets
 conf.contribs['CAN'] = {'swap-bytes': False}
 
+CAN_FRAME_SIZE = 16
```
Comment (guedou) [inline scapy/layers/can.py]: This is not needed in this PR. Please remove it.

File: scapy/layers/can.py [hunk 2]
```diff
@@ -32,11 +33,9 @@ class CAN(Packet):
     """
     fields_desc = [
         FlagsField('flags', 0, 3, ['error', 'remote_transmission_request', 'extended']),
-        XBitField('identifier', 0, 29),
-        FieldLenField('length', None, length_of='data', fmt='B'),
-        ThreeBytesField('reserved', 0),
-        StrLenField('data', '', length_from=lambda pkt: pkt.length),
-        StrLenField('padding', '', length_from=lambda pkt: 8 - pkt.length),
+        XBitField("identifier", 0, 29),
+        PadField(FieldLenField("length", None, length_of="data", fmt="B"), 4),
```
Comment (guedou) [inline scapy/layers/can.py]: Could you explain why theses changes are needed? It does not seem useful.

**Comment by polybassa** [comment]
Hi @guedou,

in the Linux Kernel, can frames are implemented with this structure:
```
define CAN_MAX_DLEN 8
...
/**
 * struct can_frame - basic CAN frame structure
 * @can_id:  CAN ID of the frame and CAN_*_FLAG flags, see canid_t definition
 * @can_dlc: frame payload length in byte (0 .. 8) aka data length code
 *           N.B. the DLC field from ISO 11898-1 Chapter 8.4.2.3 has a 1:1
 *           mapping of the 'data length code' to the real payload length
 * @__pad:   padding
 * @__res0:  reserved / padding
 * @__res1:  reserved / padding
 * @data:    CAN frame payload (up to 8 byte)
 */
struct can_frame {
	canid_t can_id;  /* 32 bit CAN_ID + EFF/RTR/ERR flags */
	__u8    can_dlc; /* frame payload length in byte (0 .. CAN_MAX_DLEN) */
	__u8    __pad;   /* padding */
	__u8    __res0;  /* reserved / padding */
	__u8    __res1;  /* reserved / padding */
	__u8    data[CAN_MAX_DLEN] __attribute__((aligned(8)));
};
```
See https://github.com/torvalds/linux/blob/master/include/uapi/linux/can.h for more details.

The current implementation produces the following output:
```
>>> load_layer("can")
>>> bytes(CAN(flags="error", identifier=1234, data="test"))
b' \x00\x04\xd2\x04\x00\x00\x00test'
```

The new implementation from this PR will produce this output:
```
>>> load_layer("can")
>>> bytes(CAN(flags="error", identifier=1234, data="test"))
b' \x00\x04\xd2\x04\x00\x00\x00test\x00\x00\x00\x00'
```

- My CANSocket implementation will directly work with Linux CAN Sockets. Therefore I need to send always 16 bytes when I do a socket.write. This is guaranteed with the new implementation. Otherwise I will have to write padding code in my socket implementation. In my opinion, the padding should be handled by the scapy CAN(Packet) implementation and not by the CANSocket.
- In my opinion, the new implementation is closer to the implementation in the Linux Kernel
- Previously we already discussed the byteswap topic. I will also do a byteswap in my CAN Socket implementation, to preserve the pcap byteorder inside scapy. This will lead to a more usable CAN implementation. For this byteswap, I'll use the following instruction ```struct.pack('<I12s', *struct.unpack('>I12s', bytes(CAN(....))))```. This instruction will also fail on the current implementation, since the length of the output of bytes is not 16. 

Greetings Nils

**Comment by guedou** [comment]
I had a look at various sources discussing CAN frame format and did not read anything related to a minimum 16 bytes length or mandatory padding. Consequently, and I might be wrong, it seems that this PR add Linux specific changes and should not be merged.

As you wrote, the padding must be done in the CAN Socket implementation. This will keep the CAN frame build and dissection work fine on other OS. I will suggest that you add a Padding field when returning the from the socket.

I am fine with merging a Linux-only CAN Socket but I don't want to change other part of the code if not necessary.

**Comment by polybassa** [comment]
Hi @guedou,
thanks for your reply. There is one example, where I'm concerend about the correct scapy like way.
Let's look at this example:
```
###(006)=[failed] CAN Socket send recv small packet

>>> def sender():
...     sleep(0.1)
...     sock2 = CANSocket(iface="vcan0")
...     sock2.send(CAN(identifier=0x7ff,length=1,data=b'\x01'))
... 
>>> thread = threading.Thread(target=sender)
>>> thread.start()
>>> 
>>> rx = sock1.recv()
>>> rx == CAN(identifier=0x7ff,length=1,data=b'\x01')
False
>>> print(bytes(rx))
b'\x00\x00\x07\xff\x01\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00'
>>> print(bytes(CAN(identifier=0x7ff,length=1,data=b'\x01')))
b'\x00\x00\x07\xff\x01\x00\x00\x00\x01'
>>> print(rx.data)
b'\x01'
>>> print(rx.padding)
b'\x00\x00\x00\x00\x00\x00\x00'
```
Since the current implementation is using a StrLenField for padding, this padding is part of the compare. A CAN packet received from a socket will always have a padding field. This means, all CAN packets, created in software which have less than 16 bytes, will differ from the equivalent packets read from a socket. 
In my opinion, this is an unwanted behaviour of scapy, and a PadField should be used in the field_desc of a CAN packet.

**Comment by guedou** [comment]
In this case, recv() should return:
```
CAN(identifier=0x7ff,length=1,data=b'\x01')/Padding(b'\x00\x00\x00\x00\x00\x00\x00')
```
Padding sometimes happens on other layer2.

**Comment by polybassa** [comment]
Alright, I'm fine with this solution. Then, the padding field of the current CAN field_desc should be removed. I don't see any use for this field. It's also not part of the pcap definition: http://www.tcpdump.org/linktypes/LINKTYPE_CAN_SOCKETCAN.html

**Comment by smainand** [comment]
I pushed a PR (#1229) that tries to perform all the modifications we discussed here.
@polybassa does it suits your needs for the socket implementation ?

**Comment by polybassa** [comment]
Hi,
Yes, this changes would make the CAN packet implementation more clear and consistent. Thanks for this changes. With this PR, I can continue with my CANSocket implementation.

**Pull request closed without merging**